bluetooth: remove blocking operation in bt_conn_get_info

liuX10 opened the pull request:
bt_conn_get_info API is used to retrieve connection-related information. However, bt_conn_get_info sends the HCI command BT_HCI_OP_READ_ENCRYPTION_KEY_SIZE to retrieve current key_size, causing excessive blocking time.

Thalley commented on subsys/bluetooth/host/keys.h:
```diff
@@ -218,6 +218,7 @@ enum {
 struct bt_keys_link_key {
 	bt_addr_t addr;
 	uint8_t storage_start[0] __aligned(sizeof(void *));
+	uint8_t key_size;
```
Correct me if I'm wrong, but wouldn't this affect `BT_KEYS_LINK_KEY_STORAGE_LEN`, and thus this change would basically invalidate/break any currently stored keys in persistent storage?

Thalley replied:
> With the changes in this PR, the key size does not need to be stored in persistent storage, as it is only used to store the encrypted key size.

In that case, I think the `key_size` should be moved above `storage_start`

> While, for the follow-up, if there is a scenario where the previously encrypted key size needs to be used, we will also encounter the situation where BT_KEYS_LINK_KEY_STORAGE_LEN is modified. For this case, how it is processed in Zephyr?

Not sure I follow this - Can you elaborate?

liuX10 replied:
BREDR's key size is read in real-time from the controller after connection establishment and encryption changes. So I think the key size need not to be stored in storage, thereby avoiding loading invalid keys.

lylezhu2012 replied:
> But when we reboot and re-enable BT, we perform an HCI Reset. How would the controller know what the key size for a specific connection is after a HCI Reset? The key size has to be stored in persistent memory if we want it to be correct after rebooting, doesn't it?

The enc_key_size of BR depends on the controller. It is different from LE. For BR, it can be got from the event `HCI_Encryption_Change[v2]`. Or it can be read by the command `HCI_Read_Encryption_Key_Size`. 
And for the command `HCI_Link_Key_Request_Reply`, the key size is not a parameter when setting up the encryption.
So I think for BR, the enc_size should not be persistent. It should be read from controller for each event of encryption changed or key refreshed.

lylezhu2012 replied:
The link key is always 128bits. And the encryption key size is negotiated. And the range is 8bits ~ 128 bits. So the encryption key size only can be got when ACL connection is established.

For I think the enc_key_size is 0 when calling `get_info`, if the ACL is not encrypted. And if the encrypt changed or key refreshed event notified, the enc key size can be got from `HCI_Encryption_Change[v2]` or `HCI_Read_Encryption_Key_Size`. And it is kept into `struct bt_keys::enc_size`. 
And the enc_key_size is `struct bt_keys::enc_size` when calling `get_info` if the ACL is encrypted.

Thalley replied:
> The link key is always 128bits. And the encryption key size is negotiated. And the range is 8bits ~ 128 bits. So the encryption key size only can be got when ACL connection is established.

Ah, given the naming, I thought `bt_keys_link_key.key_size` referred to the link key size, but it sounds like it refers to the encryption key size (which makes sense). If the link key size is always 128 bits, then I think everything here makes sense :) But perhaps `bt_keys_link_key.key_size` should be updated to `enc_key_size` to avoid confusion (for people like me who are not experts in BT Classic) - Perhaps with an additional comment that the key size will be updated once the link is encrypted, and is otherwise 0. 

One followup question: With the same link key, can the encryption key size change between connections? i.e. can the following happen:

1. Connect 2 devices. Encryption key size is 80 bits. 
2. Devices disconnect
3. Connect the same 2 devices. Encryption key size is 88 bits

Just wondering if `conn->br.link_key->key_size` should be set to `0` on disconnect

lylezhu2012 replied:
> Perhaps with an additional comment that the key size will be updated once the link is encrypted, and is otherwise 0.

Yes, I think so. 

> 1. Connect 2 devices. Encryption key size is 80 bits.
> 2. Devices disconnect
> 3. Connect the same 2 devices. Encryption key size is 88 bits

I did not notice that. I think it depends on the implementation of controller.

> Just wondering if `conn->br.link_key->key_size` should be set to `0` on disconnect

Yes, I think it is reasonable.


@liuX10 Could you please update the PR to address these comments?

Thalley replied:
Sounds like we are in agreement

So add comment/rename to make it clear that it is the size of the encryption key, and not the link key, and reset the encryption key size to 0 on disconnect, as it may change between connections, and storing the encryption key size from the last connection may be confusing

liuX10 commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1080,6 +1080,7 @@ static void hci_disconn_complete(struct net_buf *buf)
 		if (conn->type == BT_CONN_TYPE_BR && conn->br.link_key != NULL &&
 		    atomic_test_and_clear_bit(conn->flags, BT_CONN_BR_NOBOND)) {
```
@Thalley I found that there is an additional condition in this conditional statement that may prevent `enc_key_size` from being set to 0.

Thalley replied:
Hmm, my understanding from https://github.com/zephyrproject-rtos/zephyr/pull/95600#discussion_r2348466401 was that the encryption key size could be different between connections, even when we are bonded. 

i.e. even if we are bonded, we should set the encryption key size to 0, which the current patch you just pushed seemingly doesn't

liuX10 replied:
@Thalley  guess you might have misunderstood the meaning of `BT_CONN_BR_NOBOND`. This flag means pair but not bond.
Previous changes may still exist. In bond mode, `enc_key_size` is not reset to 0 as disconnect; it is only set to 0 in not bond mode alongside `Linkkey clear`.
```c
if (conn->type == BT_CONN_TYPE_BR && conn->br.link_key != NULL &&
		    atomic_test_and_clear_bit(conn->flags, BT_CONN_BR_NOBOND)) 
```
```c
now, current push ensures that the reset of `enc_key_size` is no longer tied to BOND mode or non-BOND mode.
if (conn->type == BT_CONN_TYPE_BR && conn->br.link_key != NULL) {
			if (atomic_test_and_clear_bit(conn->flags, BT_CONN_BR_NOBOND))
				bt_keys_link_key_clear(conn->br.link_key);

			conn->br.link_key->enc_key_size = 0;
}
```

Thalley reviewed (commented):
LGTM, but please fix compliance issue :)

lylezhu2012 commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2481,35 +2481,12 @@ uint8_t bt_conn_enc_key_size(const struct bt_conn *conn)
 		return 0;
 	}
 
+#if defined(CONFIG_BT_CLASSIC)
 	if (IS_ENABLED(CONFIG_BT_CLASSIC) &&
 	    conn->type == BT_CONN_TYPE_BR) {
```
```suggestion
#if defined(CONFIG_BT_CLASSIC)
	if (conn->type == BT_CONN_TYPE_BR) {
```

liuX10 commented:
hi! @jhedberg 
twister-build-prep check fail. What issue caused this?

jhedberg commented:
> hi! @jhedberg twister-build-prep check fail. What issue caused this?

This was discussed yesterday in #ci on Discord: https://discord.com/channels/720317445772017664/1014241011989487716/1420717033233977345

Solution: please rebase on latest `main`

liuX10 committed:
bluetooth: remove blocking operation in bt_conn_get_info

bt_conn_get_info API is used to retrieve connection-related
information. However, bt_conn_get_info sends the HCI command
BT_HCI_OP_READ_ENCRYPTION_KEY_SIZE to retrieve current
key_size, causing excessive blocking time.

Signed-off-by: Xiang Liu <liuxiang18@xiaomi.com>
